packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -530,9 +530,10 @@ class RenderParagraph extends RenderBox
       // Only constrain the width to the maximum width of the paragraph.
       // Leave height unconstrained, which will overflow if expanded past.
       child.layout(
-        BoxConstraints(
-          maxWidth: constraints.maxWidth,
-        ),
+        // The content will be enlarged by textScaleFactor during painting phase.
+        // We reduce contraint by textScaleFactor so that the content will fit
+        // into the box once it is enlarged.
+        BoxConstraints(maxWidth: constraints.maxWidth) / textScaleFactor,
```

Comment:
@GaryQian I wonder why do we need _textPainter.inlinePlaceholderScales? aren't they always the same as textScaleFactor?

Reply:
I believe that in actual use cases, it is indeed true, but it is also possible (at lower levels) to independently set them. This may or may not be necessary at this level, I can look into it to see if it is safe to remove.

Reply:
The scales are added into the list through WidgetSpan.build() which gets its value from the textScaleFactor. I couldn't find a possible way to have a different textscalefactor.